src/Mvc/Mvc.Core/src/Routing/DynamicRouteValueTransformer.cs

```diff
@@ -0,0 +1,14 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Routing;
+
+namespace Microsoft.AspNetCore.Mvc.Routing
+{
+    public abstract class DynamicRouteValueTransformer
+    {
+        public abstract Task<RouteValueDictionary> TransformAsync(HttpContext httpContext, RouteValueDictionary values);
```

Comment:
This is what I think encapsulates the really common set of requirements for those that subclass `Route`. You want to turn one set of route values into another one, and then let MVC do its thing.